### PR TITLE
fixing broken lake tiles in offline #335 and addresses #347

### LINKED ISSUE #335: facilitate output of potential evaporation directly from the offline code base
There are many applications where direct knowledge of potential evaporation (and/or its relation to simulated evaporation or evapotranspiration) would be of interest. However this is not facilitated by CABLE despite the `canopy%epot` variable carrying this information.  While an approximation of average potential evaporation could be estimated from other outputs, the nonlinearity in the equations implies that it would be better evaluated by CABLE's inbuilt time averaging.

Note that `ssnow%potev` in CABLE is more strictly the 'latent heat flux consistent with potential evaporation from the soil' not evaporation itself - i.e. units are currently W/m2 not mm of water/s - this may need to be handled within this issue.

### LINKED ISSUE #341: lake tiles in offline appear broken
Conceptually the lake tile in CABLE is modelled as a permanently saturated bare soil surface.  In the coupled model this is achieved in two steps:
- at the beginning of each time step (explicit section) the soil is set to saturation. any increment of water necessary to achieve this is diagnosed into `%wb_lake` (see [um_update_soilsnow](https://github.com/CABLE-LSM/ACCESS3-JULES/blob/main/src/control/cable/util/update/cbl_um_update_soilsnow.F90))
- in `soilsnow`/`surfbv` the increment `%wb_lake` is used to extract water from runoff, drainage and from the deepest soil layer in an attempt to balance water within the time step.  Any residual imbalance is retained and then used as a scaling on river outflow (in fully-coupled model).

In offline we have retained the section in `surfbv` but do not have an obvious equivalent section to the resetting of soil moisture to saturation and evaluation of `%wb_lake`.  Since `%wb_lake` is initialised to zero the whole code becomes effectively inactive.   The net result is that lake tile behaves like a bare soil tile with notable variations in soil moisture (as seen in recent analysis of the offline-as-ACCESS configuration)

Also: follow up on what is supposed to happen with wetlands - are these supposed to be saturated grasslands?

### PR DESCRIPTION
**Pull request by har917**
# CABLE

## Description

This PR aims to achieve two tasks
- removing the hard-wired lake index from the offline code (including GW module)
- maintaining lakes at saturation in offline model (soil-snow only)

Fixes #341
Addresses #347

## Type of change

- [x] Bug fix
- [x] Science advancement 

## Checklist

- [x] The new content is accessible and located in the appropriate section.
- [ ] I have checked that links are valid and point to the intended content.
- [ ] I have checked my code/text and corrected any misspellings

Please add a reviewer when ready for review.

----
📚 Documentation preview 📚: https://cable--348.org.readthedocs.build/en/348/

### CODE REVIEW

**Comment by har917** [comment]
commit 4aaccb5054aa5459c7c2d5ec97676f5fa4aa9241 produces bitwise equivalence on a case where the tumbarumba test case has been converted to a lake with very small canopy height (hc=0.001)
index_fix_Qh
index_fix_Qle
index_fix_Tscrn

**Comment by har917** [comment]
@ccarouge @JhanSrbinovsky @rml599gh @bibivking - in implementing this fix to the offline code I've noticed that the coupled model sets the lake tile water balance at field capacity (`%sfc`) and not at saturation (`%ssat`).  Can anyone advise which would be correct? (I think it should be saturation but I'm not a hydrologist)

**Comment by har917** [comment]
commit b326199e9dcbeabb54c84625949f537662b7b537 enforces saturation at the start of the time step.

The original time series of soil temperature and moisture were
original_soilM
original_soilT

This is revised to 
fixed_saturation_soilM
fixed_saturation_soilT

A comparison of original vs revised soil layer 1 moisture is
fixed_saturation_soilM1
noting that this is generated from output at end of time step so soil layer 1 can drop below `%ssat=0.42` in this output.

**Comment by har917** [comment]
The revision also impacts the surface energy balance
fixed_saturation_Qh
fixed_saturation_Qle
and diagnostic screen level temperature and humidity
fixed_saturation_Tscrn
fixed_saturation_Qscrn

The causal relationship for some of these differences is difficult to assign - there is the immediate impact of maintaining saturation (more latent, less sensible),  an immediate impact through the lower surface temperatures (less latent, less sensible and less outgoing longwave), and a delayed response because of the temporal dynamics of soil moisture and temperature (in particular the timing of the transition from <> freezing changes)

File: src/science/soilsnow/cbl_surfbv.F90 [hunk 1]
```diff
@@ -10,6 +10,8 @@ SUBROUTINE surfbv (dels, met, ssnow, soil, veg, canopy )
 
 USE smoisturev_mod,               ONLY: smoisturev
     USE cable_common_module
+    USE grid_constants_mod_cbl, ONLY : lakes_cable
```
Comment (JhanSrbinovsky) [inline src/science/soilsnow/cbl_surfbv.F90:13]: My only comment with this is that we've just pivoted in AM3 to using a namelist input to define lakes_cable. Something which will need to be ported to offline anyway - so the is a fair solution until that happens

**Comment by har917** [comment]
I am a bit concerned about the impact on the deepest soil layer moisture seen above - this is because of the lines in `surfbv` that remove water from that layer to fulfill the water requirements of the surface layer.  For this particular case this removal accumulates through time faster than the water can be replenished by downwards transport through the soil column and deep soil layer moisture reaches `(%sfc+%swilt)/2` for large parts of the year.

Note that this removal does not respect the `frozen_limit` nor any frozen fraction of soil moisture implying that the code may not conserve energy (and that the code is inefficient in that lines 112-115 is redundant given line 116-119)

Obvious tests would be
- remove 116-119 (and use the higher limit)
- remove 112-119 (and not allow water to be extracted from the deepest soil layer)
and/or
- modify these lines to acknowledge that only liquid water can be extracted and that `frozen_limit` must be abided by.

**Comment by har917** [comment]
Looking to separate concerns - will focus this issue on implementing the ***same*** physics as in coupled mode.  We will leave the investigation as to whether to use `%sfc` or `%ssat` in formulations until later - noting that this may need additional flexibility to choice of formulation of `%wetfac`.

**Comment by har917** [comment]
Following the coupled model approach and reverting to using `%sfc` in the code results in
fixed_field_soilM
fixed_field_soilT!
fixed_field_soilM1

The revised fix still extracts substantive water (over time) from the lowest layer but the gradient in soil moisture is reduce (which should have less impact on the dynamics).

**Comment by har917** [comment]
The corresponding (using `%sfc` not `%ssat`) impacts on the variables are
fixed_saturation_Qh
fixed_saturation_Qle
fixed_field_Tscrn
fixed_saturation_Qscrn

There is a much reduced impact on screen level temperature  (which is closer to expectation) using this approach.

**Comment by har917** [comment]
Benchcab tests ran successfully - noting that this is a largely meaningless assessment given that none of the sites should encounter the revised code.  These required the `check%ranges` patch - 
[config.yaml.txt](https://github.com/user-attachments/files/16370826/config.yaml.txt)

[benchmark_cable_qsub.sh.o121637899.txt](https://github.com/user-attachments/files/16370802/benchmark_cable_qsub.sh.o121637899.txt)

me.org is still down so sending to review.

**Comment by har917** [comment]
In the current `gw_hydro` code there are some conditions on `%iveg .GE. 16` which implies lakes or ice.  The current implementation removes the hard-wired indexes (replacing `16` with `lakes_cable`) but retains the `.GE.` condition (i.e. the assumed tile ordering).

It should be possible to implement a minor variation to the changeset here to use `(veg%iveg == lakes_cable) .OR. (veg%iveg == ice_cable)` as the condition and resolve this.

**Comment by JhanSrbinovsky** [comment]
sounds like a solid fix

**Comment by ccarouge** [comment]
@har917 Sorry I'm getting to this a bit late. I'm a bit puzzled with the code in `surfbv`. You say some of it is redundant. But that's not how I read it.

I read this code for lake in this way:

```
       xxx = MAX(0.0_r_2, (ssnow%wb(:,ms) - REAL(soil%sfc(:),r_2))*soil%zse(ms)*1000.0)
       ssnow%sinfil  = MIN( REAL(xxx), ssnow%wb_lake )
       ssnow%wb(:,ms) = ssnow%wb(:,ms) - REAL(ssnow%sinfil / (soil%zse(ms)*1000.0), r_2)
```
will potentially remove some water from the 6th layer and then

```
       xxx = MAX(0.0_r_2, (ssnow%wb(:,ms) - 0.5*(soil%sfc + soil%swilt))*soil%zse(ms)*1000.0)
       ssnow%sinfil  = MIN( REAL(xxx), ssnow%wb_lake )
       ssnow%wb(:,ms) = ssnow%wb(:,ms) - ssnow%sinfil / (soil%zse(ms)*1000.0)
```
will potentially remove more water.

So I don't see these as being redundant. Maybe in reality, there will never be enough water to draw twice and one of this will be 0 but there is nothing in the code stopping it to happen.

Is the logic here trying to get `wb` down to `sfc` or `0.5*(sfc+swilt)` depending on which is smaller and some restrictions according to `wb_lake` value?

**Unit issue?**
I'm also not sure about units here. If we take the first case and wb > sfc, then we have:
```
xxx = wb - sfc*zse*1000
sinfil = wb - sfc*zse*1000.    (assumption on the min here)
wb = wb - sinfil/(zse*1000) = wb - (wb -sfc*zse*1000)/(zse*1000) = wb - wb/(zse*1000)+sfc
```
This "wb-wb/(zse*1000)" is very weird., I'm not sure the division by `zse*1000` should be in there when we update `wb`. 

I'm not sure this should be part of this fix or come as a different issue or if I'm completely wrong, so looking for your input first.

**Comment by har917** [comment]
@ccarouge A couple of quick replies:
- I don't see why we need the first 
```
       xxx = MAX(0.0_r_2, (ssnow%wb(:,ms) - REAL(soil%sfc(:),r_2))*soil%zse(ms)*1000.0)
       ssnow%sinfil  = MIN( REAL(xxx), ssnow%wb_lake )
       ssnow%wb(:,ms) = ssnow%wb(:,ms) - REAL(ssnow%sinfil / (soil%zse(ms)*1000.0), r_2)
```
if we then go and do the second extraction.  Since `swilt<=sfc` always, `0.5*(sfc+swilt)<=sfc` and the net effect with/without these first 4 lines will be the same.  The code removes as much water from the 6th soil layer as required to satisfy `%wb_lake` without forcing `%wb` below `0.5*(sfc+swilt)` either way; these four lines just force it to do it in 2 steps not 1.

- the factor 1000 here is a `density_liq`.  We need to conserve water while noting that `sinfil` and `wb_lake` are in kg/m2 whereas `wb` is in m3 water/m3 soil (and `sinfil` in particular has other dependencies).  You're correct that the conversions are meaningless if you're at the limit on `xxx`, but they're not if you're not at the limit.

- there is a problem here in that none of this respects any frozen water in the soil (i.e. lake) or the frozen limit - but that's a separate issue and highly unlikely to be encountered.

**Review by ccarouge — APPROVED** [review]
All good.

**Comment by har917** [comment]
@ccarouge one final comment on units:  you said you were concerned about units in
```
xxx = wb - sfc*zse*1000
sinfil = wb - sfc*zse*1000.    (assumption on the min here)
wb = wb - sinfil/(zse*1000) = wb - (wb -sfc*zse*1000)/(zse*1000) = wb - wb/(zse*1000)+sfc
```

You missed a bracket - it should be
```
xxx = (wb - sfc)*zse*1000
sinfil = (wb - sfc)*zse*1000.    (assumption on the min here)
wb = wb - sinfil/(zse*1000) = wb - ((wb - sfc)*zse*1000)/(zse*1000) = sfc
```
so everything works out

**Pull request merged**